### PR TITLE
Balance modelling and symbolic sender variables

### PR DESCRIPTION
**Pull request by JoranHonig**
This pr implements three elements
 - Balance modelling
 - LaserEVM account interface refactor
   - This is necessary as it allows us to control the accounts that are added to a worldstate
   - Also, the new interface should be more intuitive
 - Symbolic senders

This is a more correct model than what we currently use. As a consequence of this pr we should refactor the ethersend and thief modules to work with this precise balance model.

### CODE REVIEW

File: mythril/laser/ethereum/state/account.py [hunk 1]
```diff
@@ -91,36 +92,52 @@ def __init__(
...
         :param balance:
         """
-        self.balance = balance
+        balance = (
+            symbol_factory.BitVecVal(balance, 256)
+            if isinstance(balance, int)
+            else balance
+        )
+        if self._balances is None:
+            assert False
```
Comment (norhh) [inline mythril/laser/ethereum/state/account.py]: something like `assert self._balances is not None` might be better here.

File: mythril/laser/ethereum/state/account.py [hunk 2]
```diff
@@ -91,36 +92,52 @@ def __init__(
...
-        self.balance += balance
+        balance = (
+            symbol_factory.BitVecVal(balance, 256)
+            if isinstance(balance, int)
+            else balance
+        )
+        self._balances[self.address] += balance
 
     @property
     def as_dict(self) -> Dict:
```
Comment (norhh) [inline mythril/laser/ethereum/state/account.py]: the parentheses of self.balance are missing here.
it should be `"balance": self.balance()`

File: mythril/laser/ethereum/state/world_state.py [hunk 3]
```diff
@@ -22,18 +24,29 @@ def __init__(
         :param transaction_sequence:
         :param annotations:
         """
-        self.accounts = {}  # type: Dict[str, Account]
+        self._accounts = {}  # type: Dict[int, Account]
+        self.balance = Array("balance", 256, 256)
```
Comment (norhh) [inline mythril/laser/ethereum/state/world_state.py]: Any reason why this is singular? (self.balance over self.balances)

File: mythril/laser/ethereum/state/world_state.py [hunk 4]
```diff
@@ -22,18 +24,29 @@ def __init__(
         :param transaction_sequence:
```
Comment (norhh) [inline mythril/laser/ethereum/state/world_state.py]: The docstrings for the new variables.

File: mythril/laser/ethereum/transaction/symbolic.py [hunk 5]
```diff
@@ -69,12 +78,15 @@ def execute_contract_creation(
     :return:
     """
     # TODO: Resolve circular import between .transaction and ..svm to import LaserEVM here
-    open_states = laser_evm.open_states[:]
+    # open_states = laser_evm.open_states[:]
```
Comment (norhh) [inline mythril/laser/ethereum/transaction/symbolic.py]: Can this be removed?

File: mythril/laser/smt/bool.py [hunk 6]
```diff
@@ -90,15 +90,18 @@ def And(*args: Union[Bool, bool]) -> Bool:
...
 
     :param a:
     :param b:
     :return:
     """
-    union = a.annotations + b.annotations
-    return Bool(z3.Or(a.raw, b.raw), annotations=union)
+    union = []
+    args_list = [arg if isinstance(arg, Bool) else Bool(arg) for arg in args]
+    for arg in args_list:
```
Comment (norhh) [inline mythril/laser/smt/bool.py]: We can also combine all the 3 lines to `union = [arg.annotations for arg in args_list]` which might be a 1 liner

File: mythril/laser/ethereum/svm.py [hunk 7]
```diff
@@ -106,35 +98,44 @@ def __init__(
...
-            log.info("Starting message call transaction to {}".format(main_address))
-            self._execute_transactions(main_address)
+        if pre_configuration_mode:
+            self.open_states = [world_state]
+            log.info("Starting message call transaction to {}".format(target_address))
+            self._execute_transactions(symbol_factory.BitVecVal(target_address, 256))
 
-        elif creation_code:
+        elif scratch_mode:
+            # self.open_states = [WorldState()]
```
Comment (norhh) [inline mythril/laser/ethereum/svm.py]: any reason why this is commented?

File: mythril/laser/ethereum/state/account.py (hunk 1 above)
Comment (JoranHonig) [reply mythril/laser/ethereum/state/account.py]: This was probably here like this to debug

Comment (JoranHonig) [reply mythril/laser/ethereum/state/account.py]: Fixed now 👍

File: mythril/laser/ethereum/state/account.py (hunk 2 above)
Comment (JoranHonig) [reply mythril/laser/ethereum/state/account.py]: Good catch

File: mythril/laser/ethereum/state/world_state.py (hunk 3 above)
Comment (JoranHonig) [reply mythril/laser/ethereum/state/world_state.py]: not really, I changed it to be plural 👍

File: mythril/laser/ethereum/svm.py (hunk 7 above)
Comment (JoranHonig) [reply mythril/laser/ethereum/svm.py]: Yeah, execute contract creation set's up the global state

File: mythril/laser/ethereum/transaction/symbolic.py (hunk 5 above)
Comment (JoranHonig) [reply mythril/laser/ethereum/transaction/symbolic.py]: yes

File: mythril/laser/ethereum/svm.py (hunk 7 above)
Comment (norhh) [reply mythril/laser/ethereum/svm.py]: I mean this can be removed right?

Comment (JoranHonig) [reply mythril/laser/ethereum/svm.py]: it is 😅

**Comment by JoranHonig** [comment]
@norhh is this good to go like this?

**Pull request merged**